Fix part of #6550: Increased coverage for learner progress services

EricZLou opened the pull request:
## Explanation
Fixes part of #6550 for writing backend tests for learner_progress_services. Deletes unused function in learner_progress_services and adds functionality in learner_progress_services_test to test if an exploration or collection is not public.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

lou-eric committed:
100% coverage for learner progress services

kevinlee12 commented:
Hi @ezl-13, there are some lint tests that are failing, ptal!

EricZLou commented:
Thanks @kevinlee12 ! 
Lint tests are telling me not to use "\" to continue code from one line to the next when a line is too long to fit under 80 characters. How else should I do this especially if the statement doesn't have parentheses (it has "=")? 
Also, I'm getting the error "Access to a protected member _get_filtered_completed_exp_summaries of a client class (protected-access)". It is true that this function is from learner_progress_services (which I guess is protected). But this error doesn't come up when calling the function mark_exploration_as_completed which is also from learner_progress_services. How can I get around this?

lilithxxx commented:
> Lint tests are telling me not to use "" to continue code from one line to the next when a line is too long to fit under 80 characters. How else should I do this especially if the statement doesn't have parentheses (it has "=")?

Hi @ezl-13 You can use `# pylint: disable=line-too-long` but use it only if there is no other way to break the line. 

> Also, I'm getting the error "Access to a protected member _get_filtered_completed_exp_summaries of a client class (protected-access)". It is true that this function is from learner_progress_services (which I guess is protected). But this error doesn't come up when calling the function mark_exploration_as_completed which is also from learner_progress_services. How can I get around this?

You should not access the private methods but instead test them through a public method as written in the 9th point [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests)

EricZLou commented:
@lilithxxx thanks so much for the help! I'll work on that now.

EricZLou commented:
I've been trying to find a way to test it through a public interface, and I was wondering if it is possible to update exploration summaries. For example, the status of an exploration summary is usually ACTIVITY_STATUS_PUBLIC. I want to change it to ACTIVITY_STATUS_PRIVATE. Is this possible?

lilithxxx commented:
Hm maybe use `self.swap()`? Try to look at other files and see how the status of the exploration summaries are changed. Does this help? Thanks

seanlip commented:
Hi @lilithxxx, can you elaborate on how exactly to use self.swap() here? I'm not sure I understand your suggestion.

EricZLou commented:
@lilithxxx I did a search (on "ExplorationSummary status") to see if there were any places in the repo where the status of an ExplorationSummary object is changed, and I believe it is never changed. I also searched for "ACTIVITY_STATUS_PRIVATE" and none of them are related to an ExplorationSummary status. 
Perhaps the untested lines in learner_progress_services "elif exploration_summary.status != constants.ACTIVITY_STATUS_PUBLIC:" were precautionary checks, though they are impossible to happen since the status is never changed to a non-public status?

seanlip commented:
@ezl-13 don't explorations start off as private (and then only turn public once their creator publishes them)?

seanlip commented:
(Also note that explorations can be unpublished by a moderator/admin after they've been published.)

lilithxxx commented:
Sorry @seanlip I didn't read through the code -- it was a general solution.
@ezl-13 I think by unpublishing the explorations, they become private

EricZLou commented:
Thanks a lot for the guidance @seanlip @lilithxxx .
I think I have it down now through unpublishing.
In the 6 functions that require testing for ACTIVITY_STATUS_PRIVATE, the code I am adding is nearly identical, so I created a new test function that takes in a parameter that each of the 6 functions can call. 
However, I realize that test functions can only take in self as a parameter. Any tips on what to do here?
I would like to make a function prototype like this in the test file: 

def test_private_activity(self, category, specific_type):

The long solution is, of course, to not try to compress the code and have nearly identical code for all 6 functions. Do I go with this?

lou-eric committed:
Successful testing of Private. Incorrect function prototype.

EricZLou commented:
Just a note that the most recent commit was just for my own record (to keep track of my progress). It definitely won't pass tests for the reason that I commented above.

DubeySandeep commented:
@ezl-13, I think you can create a separate helper function for this one and helper function should help to create a scenario (and maybe testing some general behavior too.)  but the main testing should be done on each separate test function. Does this help?

EricZLou committed:
Helper functions for private activity

EricZLou commented:
Fixing the linting errors right now!

EricZLou committed:
Fix linting Errors

EricZLou commented:
I'm currently getting this linting error that I'm not sure how to fix. Any recs? 
************* Module oppia.schema_utils
C:214,15: More than one statement on a single line (multiple-statements)

DubeySandeep commented:
@ezl-13, Can you please update your feature branch with upstream/develop? (This will fix the lint issue.)

EricZLou committed:
Merge branch 'develop' of https://github.com/oppia/oppia into ezl.branch

EricZLou commented:
@DubeySandeep I just made a new commit! Not sure why it's coming up as  "Merge branch 'develop' of https://github.com/oppia/oppia into ezl.branch",  but I believe I updated my feature branch with upstream/develop correctly. Let me know if I did something wrong.

EricZLou commented:
Hi @DubeySandeep @nithusha21 can someone take a look at my PR?
Thanks a lot!

DubeySandeep commented:
> @DubeySandeep I just made a new commit! Not sure why it's coming up as "Merge branch 'develop' of https://github.com/oppia/oppia into ezl.branch", but I believe I updated my feature branch with upstream/develop correctly. Let me know if I did something wrong.

@ezl-13, That's a commit message for updating the branch with the oppia/develop and it's fine! :)

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -522,6 +597,24 @@ def test_get_all_incomplete_exp_ids(self):
+        learner_progress_services.mark_exploration_as_completed(
+            self.user_id, self.EXP_ID_2)
+        exploration_ids = [self.EXP_ID_0, self.EXP_ID_1, self.EXP_ID_2]
+        exploration_summaries = exp_services.get_exploration_summaries_matching_ids( # pylint: disable=line-too-long
```
Try not to use the pylint-disable. 
```
        exploration_summaries = (
            exp_services.get_exploration_summaries_matching_ids( ...

```

nithusha21 replied:
Do this throughout the file :)

EricZLou replied:
Done :)

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -522,6 +597,24 @@ def test_get_all_incomplete_exp_ids(self):
                 self.user_id), [self.EXP_ID_0, self.EXP_ID_1])
 
+        # Mark EXP_ID_2 as completed and create summary list to test output.
+        learner_progress_services.mark_exploration_as_completed(
```
Can we create a separate test for testing this behavior? (Following the instruction provided in [the  wiki page](https://github.com/oppia/oppia/wiki/Writing-backend-tests))

EricZLou replied:
Can you clarify what you're referring to by "this behavior"? 
Do you mean test "mark_exploration_as_completed" and "mark_collection_as_completed"?

DubeySandeep replied:
> Line 611: Test that completed exp summaries don't include private exploration

The new behavior which you are testing^.

EricZLou replied:
@DubeySandeep Just to clarify, I should split up each of the 4 test functions I wrote into 2, where the FIRST one is all the code before line 611, and the SECOND one is all the code before line 611 (without the assert statements) and all the code after line 611.
Is this correct?

EricZLou replied:
Update: I understand what you were pointing out and made a new PR!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -541,6 +634,24 @@ def test_get_all_incomplete_collection_ids(self):
                 self.user_id), [self.COL_ID_0, self.COL_ID_1])
 
+        # Mark COL_ID_2 as completed and create summary list to test output.
+        learner_progress_services.mark_collection_as_completed(
```
Ditto*

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -522,6 +597,24 @@ def test_get_all_incomplete_exp_ids(self):
+
+        # Test that completed exp summaries don't include private exploration.
+        self.assertEqual(
+            incomplete_exp_summaries[0].id, exploration_summaries[0].id)
```
I'm not sure why we are testing exploration summaries here (in the learner_progress_services_test.py)? Can you please explain?

EricZLou replied:
Of course! 

We start with all the exploration_summaries of EXP_ID_0, EXP_ID_1, and EXP_ID_2. (In this block of code, they all happen to be incomplete, though this is not important). Then we pass these 3 exploration_summaries to _test_add_private_exploration_activity() which makes EXP_ID_2 private. Therefore, when we call get_activity_progress in _test_add_private_exploration_activity(), we are testing to ensure that EXP_ID_2 is not in the final list of summaries (incomplete_exp_summaries) returned by get_activity_progress (since it is private). It should, instead, contain ONLY EXP_ID_0 and EXP_ID_1.

In the 3 assert statements, we first ensure that EXP_ID_0 is in the list. Then we ensure that EXP_ID_1 is in the list. Then we ensure that the list only contains 2 items (aka EXP_ID_2 is not in it).

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -465,6 +470,40 @@ def test_remove_collection_from_incomplete_list(self):
         self.assertEqual(self._get_all_incomplete_collection_ids(
             self.user_id), [])
 
+    def _test_add_private_exploration_activity(self):
```
This function is returning something, right?  maybe you can rename this to get_users_activity_progress? 

Also, I think you can move this part directly to the tests (instead of craeting a seperate function) as this seems to be a critical path for testiong one of the functionalty.

EricZLou replied:
Will do! Yes it is returning something.

DubeySandeep replied:
@ezl-13, is this comment resolved?

EricZLou replied:
Yes. Resolved!

DubeySandeep reviewed (changes requested):
Hi @ezl-13, Thanks for adding the tests :), I've taken the first pass and left a few comments PTAL!

nithusha21 commented on core/domain/learner_progress_services_test.py:
```diff
@@ -18,13 +18,18 @@
 from core.platform import models
 from core.tests import test_utils
+import feconf
+
```
Do you need this extra line?

EricZLou replied:
I was getting a linting error for it, but I believe it was actually because import feconf should have gone after import datetime to keep the imports together. I'll fix that!

nithusha21 commented on core/domain/learner_progress_services_test.py:
```diff
@@ -465,6 +470,40 @@ def test_remove_collection_from_incomplete_list(self):
+        # Call get_activity_progress to get filtered progress.
+        user_activity = learner_progress_services.get_activity_progress(
+            self.user_id)
+        return user_activity[0]
```
Why is only the first element returned?

EricZLou replied:
get_activity_progress returns a list user_activity where:
user_activity[0] = LearnerProgress object which contains:
        filtered_incomplete_exp_summaries,
        filtered_incomplete_collection_summaries,
        filtered_completed_exp_summaries,
        filtered_completed_collection_summaries,
        filtered_exp_playlist_summaries,
        filtered_collection_playlist_summaries
user_activity[1] = map that contains number of nonexistent ids
user_activity[2] = something about completed_to_incomplete_collection_summaries (which is not relevant)

So I only need user_activity[0] to check the filtered list of summaries.

nithusha21 commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +520,24 @@ def test_get_all_completed_exp_ids(self):
+        learner_progress_services.mark_exploration_as_completed(
+            self.user_id, self.EXP_ID_2)
+        exploration_ids = [self.EXP_ID_0, self.EXP_ID_1, self.EXP_ID_2]
+        exploration_summaries = exp_services.get_exploration_summaries_matching_ids( # pylint: disable=line-too-long
```
Why do we need the exploration summaries? I think the get_all_completed_exp_ids (as used above) would suffice?

EricZLou replied:
I still do need to retrieve a list of exploration summaries (not just exploration IDs) in order to test that whatever is returned from get_activity_progress legitimately filtered the original summaries. 
I definitely can, however, change this to be 
`exp_services.get_exploration_summaries_matching_ids(
learner_progress_services.get_all_completed_exp_ids(self.user_id))`

nithusha21 reviewed (commented):
Thanks @ezl-13 for the PR. I am sorry for the delay in reviewing. Was caught up with exams. I left some comments (I left them in one of the tests, but they apply to all the other tests too, as they are similar) PTAL.

EricZLou commented:
@nithusha21 No worries at all about the delay. Focus on your exams, good luck and thank you for the feedback!!!

EricZLou commented:
@DubeySandeep please take a look at the comments. Thanks!

EricZLou committed:
Merge branch 'develop' of https://github.com/oppia/oppia into ezl.branch

EricZLou committed:
New test structure

ezl-13 commented:
(fixing the linting errors now)

EricZLou committed:
Fix linting errors of import statements

EricZLou commented:
@DubeySandeep @nithusha21 Just made a new commit! I've:
1. Split up the tests to make it more clear what behavior I'm testing.
2. Taken away lots of pylint --disables. 
Thanks for the feedback!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +485,43 @@ def test_get_all_completed_exp_ids(self):
+
+        # Test that completed exp summaries don't include private exploration.
+        self.assertEqual(
+            completed_exp_summaries[0].id, exploration_summaries[0].id)
```
Assert this with string value:
`self.assertEqual(completed_exp_summaries[0].id, <some-string-value>)`

Ditto elsewhere!

EricZLou replied:
Done!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +485,43 @@ def test_get_all_completed_exp_ids(self):
+        self.assertEqual(
+            completed_exp_summaries[1].id, exploration_summaries[1].id)
+        self.assertEqual(len(completed_exp_summaries), 2)
+
```
Can we also check what happens when the same exploration marked as "Private" before got published again? (Ditto elsewhere!)

(Trying to test the behavior.)

EricZLou replied:
Definitely can do this!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +485,43 @@ def test_get_all_completed_exp_ids(self):
+                learner_progress_services.get_all_completed_exp_ids(
+                    self.user_id)))
+
+        # Unpublish EXP_ID_2 to change status to ACTIVITY_STATUS_PRIVATE.
```
Can we assert that initially there were three exploration_ids in the list? (as a baseline verification.)

Ditto elsewhere!

EricZLou replied:
Good point. Done!

DubeySandeep commented on core/domain/learner_progress_services.py:
```diff
@@ -430,26 +430,6 @@ def _remove_activity_ids_from_playlist(
         learner_playlist_services.save_learner_playlist(learner_playlist)
 
 
-def remove_exp_from_completed_list(user_id, exploration_id):
```
Thanks for cleaning i.e, deleting the unwanted codes! :)

EricZLou replied:
:)

DubeySandeep reviewed (commented):
@ezl-13, This looks much cleaner and better now! I've left a few comments on adding a few more test and some general patterns. PTAL!

EricZLou committed:
Add tests for re-publishing private exp/col

EricZLou commented:
@DubeySandeep Finished up the edits! This is what I changed:
1. Changed assert statements at end of tests to test against string literals.
2. Added 4 tests for re-publishing a private exploration / collection, and then marking it as complete / incomplete. 
3. Assert that there were initially 3 explorations / collections in the list.

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -465,6 +469,28 @@ def test_remove_collection_from_incomplete_list(self):
         self.assertEqual(self._get_all_incomplete_collection_ids(
             self.user_id), [])
 
+
```
Remove unwanted newlines.

EricZLou replied:
Ah right. I forgot to delete those. I did that to make it more structured for myself.

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+            self.user_id, self.EXP_ID_3)
+        self.assertEqual(
+            len(learner_progress_services.get_all_completed_exp_ids(
+                self.user_id)), 3)
```
This looks great now, can we also check the ids of each exploration like you have done in line 539? (Sorry if my last comment related to this was not clear.)

EricZLou replied:
Yep, definitely!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+            self.user_id, self.EXP_ID_1)
+        learner_progress_services.mark_exploration_as_completed(
+            self.user_id, self.EXP_ID_3)
+
```
Do a baseline verification of existing completed exploration.

EricZLou replied:
Done!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+            self.EXP_ID_3)
+        self.assertEqual(
+            private_exploration.status, constants.ACTIVITY_STATUS_PRIVATE)
+
```
Ditto, assert there are only two completed exploration exist after the above steps.

EricZLou replied:
Sure.

EricZLou replied:
Actually, I cannot do this because there still are 3 completed explorations after making something private. 
Unpublishing an exploration / making it private does not change whether an exploration is completed or not. So, all 3 are still completed.

EricZLou replied:
The real differentiation comes into play when I call get_activity_progress where we will see EXP_ID_3 is no longer completed.

DubeySandeep replied:
So we expected `get_activity_progress` to return 2 explorations, can you assert this value around these lines?

EricZLou replied:
@DubeySandeep Yep, I already do this at the end of each test. I assert that either 2 or 3 explorations are filtered.

DubeySandeep replied:
@ezl-13 I don't see any assertion around these lines (After making exploration private and before publishing), have I missed something?

EricZLou replied:
@DubeySandeep Ah, I see what you mean. So each test should have 2 calls to get_activity_progress. I'll fix that!

EricZLou replied:
I feel like it would be redundant with the test above it, since I will be copy-ing and paste-ing all of it, but I will do it.

seanlip replied:
Wait, just to check, which parts are you copy-and-pasting? We should definitely avoid wholesale copy/paste so I figured it might be worth double checking.

You should probably only be adding assert statements and nothing else.

EricZLou replied:
I believe that in order to check if get_activity_progress returns only 2 explorations after marking an exploration as private, I essentially am replicating the behavior in test_get_filtered_completed_exp_summaries.

Checking that there are only 2 explorations after marking an exploration as private is exactly all the code from "# Call get_activity_progress to get filtered progress" to "self.assertEqual(len(completed_exp_summaries), 2)"

DubeySandeep replied:
> check if get_activity_progress returns only 2 explorations after marking an exploration as private, I essentially am replicating the behavior in test_get_filtered_completed_exp_summaries.

In test_get_filtered_completed_exp_summaries that test is `endline verification` whereas in this test the same thing is `baseline verification`. I thinking assertion related code will be replicated and we should avoid replicate `actions` related code.

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+
+        # Publish EXP_ID_3 to change status back to ACTIVITY_STATUS_PUBLIC.
+        self.publish_exploration(self.owner_id, self.EXP_ID_3)
+        learner_progress_services.mark_exploration_as_completed(
```
Do we again have to mark this as completed? 

I'm not sure about the implementation, but as I can see earlier this exploration was marked as "completed" so after the exploration is marked private do we remove the exploration ids from the "completed" list of each users (if they exist)? Can you please check!

EricZLou replied:
We actually don't need to. Good catch!

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+            all_filtered_summaries.completed_exp_summaries)
+
+        # Test that completed exp summaries includes all 3 of the explorations.
+        self.assertEqual(len(completed_exp_summaries), 3)
```
Assert the exploration ids (similar to line 539)

EricZLou replied:
Done :)

DubeySandeep commented on core/domain/learner_progress_services_test.py:
```diff
@@ -481,6 +507,89 @@ def test_get_all_completed_exp_ids(self):
+
+
+
+
```
Remove extra new lines.

EricZLou replied:
Done.

DubeySandeep reviewed (commented):
Hi @ezl-13, thanks for adding the new test. :) I've left a few comments for the test related to "Completed exploration" and these comments are general for all other tests "complete/incomplete exploration/collection." Can you please go through it.

Also, once these comments are addressed I'll go ahead and approve the PR. Thanks! :)

EricZLou committed:
Added some assert statements

EricZLou commented:
Hi @DubeySandeep . Made some changes, check them out please!

EricZLou commented:
@DubeySandeep Please check my reply to your comment on line 562! Thank you!

EricZLou committed:
Some changes

DubeySandeep commented:
@ezl-13, any update on this PR?

EricZLou commented:
@DubeySandeep Sorry for the delay. I haven't gotten the chance to work on it this week, but I will have some updates on Wednesday.

EricZLou committed:
updated republishing an unpublished exp/col test

EricZLou commented:
@DubeySandeep I submitted some changes including:
test names are clearer
republishing exp/col is simpler, testing with only 1 exp/col

Please take a look! Thanks.

lilithxxx commented:
@ezl-13 a couple of lines are still untested. Please make coverage 100%. Feel free to ask if you have any question, thanks!

EricZLou committed:
add tests for exploration playlist and collection playlist

EricZLou commented:
@lilithxxx thanks for the message! I believe I fixed that up. I added 4 additional tests to sufficiently cover those 2 lines. I'll check back again once the codecov report comes back!

EricZLou commented:
@DubeySandeep now the coverage should be at 100%! Please take a look. Thank you!

DubeySandeep reviewed (approved):
@ezl-13, Thanks for adding tests and making the cover 100%, LGTM! :)

@nithusha21, Can you please review this PR as a code owner?

DubeySandeep commented:
@nithusha21, can you please review this PR as a codeowner?

nithusha21 reviewed (approved):
Thanks for the PR @ezl-13 and thanks for reviewing @DubeySandeep. Sorry for the delays from my side!
